napi-inl.h

```diff
@@ -5692,6 +5709,46 @@ Addon<T>::DefineProperties(Object object,
 }
 #endif  // NAPI_VERSION > 5
 
+template <typename Hook, typename Arg>
+Env::CleanupHook<Hook, Arg> Env::AddCleanupHook(Hook hook, Arg* arg) {
+  return CleanupHook<Hook, Arg>(*this, hook, arg);
+}
+
+template <typename Hook>
+Env::CleanupHook<Hook> Env::AddCleanupHook(Hook hook) {
+  return CleanupHook<Hook>(*this, hook);
+}
+
+template <typename Hook, typename Arg>
+Env::CleanupHook<Hook, Arg>::CleanupHook(Napi::Env env, Hook hook)
+    : wrapper(Env::CleanupHook<Hook, Arg>::Wrapper) {
+  data = new CleanupData{std::move(hook), nullptr};
+  napi_status status = napi_add_env_cleanup_hook(env, wrapper, data);
+  NAPI_FATAL_IF_FAILED(status,
```

Comment:
What would cause napi_add_env_cleanup_hook to fail ?   I'm wondering if we can let the caller trying add a hook know it failed versus causing a FATAL error.

Reply:
Looking at the implementation, it looks like it should always return `napi_ok` ... https://github.com/nodejs/node/blob/663d7e9fb27f017f38b79382a95fec10e5f66f9a/src/node_api.cc#L656-L676

According to the docs it can just abort the process with invalid arguments: 
> Providing the same fun and arg values multiple times is not allowed and will lead the process to abort.

But we would never provide the same `arg` value (as we always create a `new CallbackData`)

Maybe we don't need the check at all...? However `napi_add_env_cleanup_hook` always returning `napi_ok` is an implementation detail, maybe we should keep it blackboxed 🤷 

How would you recommend "add a hook know it failed" ...?

Thanks, Kevin

Reply:
How would you recommend "add a hook know it failed" . What I had in mind was some kind of return value the caller could check. For example if it was documented that the callers should check that the result from AddCleanupHook was non-null would that make any sense?

Reply:
Addressed. Added a `bool IsEmpty()` method on `Env::CleanupHook` https://github.com/nodejs/node-addon-api/blob/36ef600b196f1876cf3390610960488e2dae9be5/napi-inl.h#L5756-L5759

`data` is nullptr if adding the hook failed:
https://github.com/nodejs/node-addon-api/blob/36ef600b196f1876cf3390610960488e2dae9be5/napi-inl.h#L5737-L5746